### ISSUE TITLE
Notepad++ Code Alignment Error - System.IO.FileNotFoundException

### ISSUE THREAD

**Issue by TCWORLD** [cpmcgrath/codealignment#47]
In the last couple of weeks I've stopped being able to use the code alignment plugin for NP++.

I get the following unhandled exception message when I try to use the align by character option:

```
System.IO.FileNotFoundException: Could not load file or assembly 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' or one of its dependencies. The system cannot find the file specified.
File name: 'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
   at System.Configuration.TypeUtil.GetTypeWithReflectionPermission(IInternalConfigHost host, String typeString, Boolean throwOnError)
   at System.Configuration.MgmtConfigurationRecord.CreateSectionGroupFactory(FactoryRecord factoryRecord)
   at System.Configuration.MgmtConfigurationRecord.EnsureSectionGroupFactory(FactoryRecord factoryRecord)
   at System.Configuration.MgmtConfigurationRecord.GetSectionGroup(String configKey)
   at System.Configuration.Configuration.GetSectionGroup(String sectionGroupName)
   at System.Configuration.ClientSettingsStore.DeclareSection(Configuration config, String sectionName)
   at System.Configuration.ClientSettingsStore.GetConfigSection(Configuration config, String sectionName, Boolean declare)
   at System.Configuration.ClientSettingsStore.WriteSettings(String sectionName, Boolean isRoaming, IDictionary newSettings)
   at System.Configuration.LocalFileSettingsProvider.SetPropertyValues(SettingsContext context, SettingsPropertyValueCollection values)
   at System.Configuration.SettingsBase.SaveCore()
   at System.Configuration.SettingsBase.Save()
   at System.Configuration.ApplicationSettingsBase.Save()
   at CMcG.CodeAlignment.FormCodeAlignment.Ok(Object sender, EventArgs e)
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.PerformClick()
   at System.Windows.Forms.Form.ProcessDialogKey(Keys keyData)
   at System.Windows.Forms.Control.ProcessDialogKey(Keys keyData)
   at System.Windows.Forms.Control.PreProcessMessage(Message& msg)
   at System.Windows.Forms.Control.PreProcessControlMessageInternal(Control target, Message& msg)
   at System.Windows.Forms.Application.ThreadContext.PreTranslateMessage(MSG& msg)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].

```

If I try to do alignment by one of the default characters like equals (the ones in the menu), it all works fine.

I'm using Windows 8.1 x64 and have the latest versions of Dot Net (4.5.1 and 4.6.1) installed and the latest versions of Notepad++ (6.9.2) and the code alignment plugin (12.0).

On another Windows 7 computer which has the same software installed, I don't have the issue.

I'm trying to work out what must have changed on my computer to make it stop working and I can't seem to figure out what has changed. The only thing I have installed since I last remember it working is the Intel OpenCL SDK, but I doubt that will be causing the issue. There haven't been any windows updates installed either, so I am really not sure what is going on.

**Comment by cpmcgrath** [comment]
Hi, Have you got .NET 3.5 installed? .NET is a funny thing, there's basically two runtime stacks the pre .NET 4 and post .NET 4. Installing .NET4.5 will let you run a .NET 4 program, but not a .NET 3.5 program.

**Comment by TCWORLD** [comment]
The odd thing is it was working perfectly fine and I'm not sure what changed. Although having gone back and looked more closely, there was a .NET 4.5.2 security update installed a month ago which could have stopped it working.

I think I must already have 3.5 installed. I tried to run the installer but it doesn't load (nothing happens when I run it). I'm not even sure if it is possible to install separately on Windows 8.1 - it seems to be part of the operating system.

**Comment by TCWORLD** [comment]
The odd thing is that Visual Studio 2013 knows that System Version 4.0.0.0 exists. I've tried downloading the repository and load the project in VS. 

As a side note, I've also tried compiling it, but I can run anything that I compile as NP++ complains that the compiled plugin is ANSI not Unicode and I can't figure out how to get it to compile as Unicode.

**Comment by cpmcgrath** [comment]
It's strange, as I said it's .NET 3.5 so it shouldn't be looking for v4 of the System assembly (It should be looking for v2.0.0.0).

As for compiling it, there's a bit of magic involved in getting it working with Npp that I don't completely understand and I've never been able to get it to working on any version of Visual Studio other than 2010. Unfortunately, the people who developed that bit of code seem to have no interest in updating it for newer versions.

It's hard to give any advice without more information. You could try uninstalling .NET 3.5 and reinstalling it (and then reinstalling code alignment) You can do this in "Programs and Features" under the "Turn Windows features on or off" option (On my machine I don't have the WCF options ticked so no need to install them).

**Comment by TCWORLD** [comment]
I've just tried that. I uninstalled 3.5, rebooted and loaded NP++. Windows detected the plugin required .NET 3.5 and asked me to install it. However after doing so I still have the same problem.

The odd thing is the public key that the error message mentions matches exactly with the System Version 2.0.0.0 that is in the C:\Windows\assemblies directory.

Really confused as to what could be causing the issue.

**Comment by TCWORLD** [comment]
Well I got it to compile in VS2013 and correctly be picked up by NP++ as Unicode. In the CodeAlignment.Npp project, I loaded the NuGet Package Manager Console and did "Install-Package UnmanagedExports". Then to fix the errors that resulted from doing that, I changed line 4 in UnmanagedExports.cs to be `using RGiesecke.DllExport` instead of `using NppPlugin.DllExport`. Now I can compile it in VS2013 and have what I compile get loaded by NP++.

Now just need to work out why it is trying to load the wrong package.

**Comment by cpmcgrath** [comment]
Wow, thanks for that! I've been meaning to remove the requirement on VS2010 to compile for a while. There are Nuget packages to build extensions for Visual Studio 2010 on more recent versions, but the annoyances with the Notepad++ build has stopped me from trying.

**Comment by TCWORLD** [comment]
I've put the vs2013 project on a branch [here](https://github.com/TCWORLD/codealignment/tree/vs-2013).

I'm still no closer to figuring out why it is trying to reference .NET V4 assemblies though.

**Comment by TCWORLD** [comment]
Well, it is definitely loading .NET 3.5 according to the process explorer. I just can't figure out why it is looking for the wrong System assembly. I'm sure my computer is haunted.

https://xkcd.com/1316/ springs to mind!

![processexplorer](https://cloud.githubusercontent.com/assets/1176680/16031132/e1118efe-31c5-11e6-897d-5c62c1b643bf.PNG)

**Comment by cpmcgrath** [comment]
In the CodeAlignment.Npp I do overload the Assembly Resolve so I can shove the other assemblies in a subfolder (Otherwise Notepad++ tries to load them as plugins)

It shouldn't affect the System assemblies but could be a good place to start. It's in the UnmanagedExports class.

**Comment by TCWORLD** [comment]
Ok, so given nothing that I have been trying seems to have done anything, I tried downloading a fresh copy NP++ again and installing it as standalone (not using APPDATA), and installed the plugin in that, and all works fine!

It then occurred to me that I had installed another plugin in the time between knowing Code Alignment works and it stopping working, a plugin which I hadn't tried on the other computer I have with the same setup. I had also since removed the plugin which is why it didn't occur to me earlier.

It seems it had left something behind in the APPDATA folder for NP++ which was causing a .NET 4 assembly to be loaded - mscoreie.dll I think it was. I've deleted the folder it left behind, and voila, Code Alignment works perfectly again!

So basically a plugin called "hwdebugger" does not happily coexist with Code Alignment. I will have to remember not to install that one again then.
